File: te_phoenix/lib/te_phoenix/npc_brain.ex
defmodule TePhoenix.NpcBrain do
  @moduledoc """
  Multi-provider AI NPC dialogue system with rule-based fallback.
  Ported from npc_brain.js. Supports Gemini, Anthropic, OpenAI, Ollama.

  Provider priority: DB config (system_settings) > env vars > rule-based fallback.
  Configure via AdminSauce > Settings > AI Brain.
  """

  require Logger

  alias TePhoenix.Repo

  @doc "Get an NPC reply. Tries AI provider, falls back to rule-based."
  def get_reply(opts) do
    prompt = build_prompt(opts)
    ai_config = opts[:ai_config] || load_ai_config()

    provider = ai_config["provider"] || ai_config[:provider]

    result = case provider do
      "gemini" -> call_provider(:gemini, prompt, ai_config)
      "anthropic" -> call_provider(:anthropic, prompt, ai_config)
      "openai" -> call_provider(:openai, prompt, ai_config)
      "ollama" -> call_provider(:ollama, prompt, ai_config)
      _ -> nil
    end

    case result do
      {:ok, text} when text != "" -> text
      _ -> rule_based_reply(opts[:npc], opts[:message], opts[:memory])
    end
  end

  @doc "Extract facts from a player message."
  def extract_facts(message, existing_facts \\ []) do
    msg = String.downcase(message)
    facts = existing_facts

    facts = if contains_any?(msg, ~w(warrior fighter sword shield)), do: add_fact(facts, "Appears to be a warrior type"), else: facts
    facts = if contains_any?(msg, ~w(mage wizard spell magic)), do: add_fact(facts, "Appears to be a mage type"), else: facts
    facts = if contains_any?(msg, ~w(rogue thief sneak steal)), do: add_fact(facts, "Appears to be a rogue type"), else: facts
    facts = if contains_any?(msg, ~w(quest job task mission)), do: add_fact(facts, "Has asked about quests"), else: facts
    facts = if contains_any?(msg, ~w(buy sell shop trade price)), do: add_fact(facts, "Is interested in trading"), else: facts
    facts = if contains_any?(msg, ~w(thanks grateful)), do: add_fact(facts, "Was polite and thankful"), else: facts
    facts = if contains_any?(msg, ~w(damn fool idiot useless)), do: add_fact(facts, "Was rude"), else: facts
    facts = if contains_any?(msg, ~w(ogham blood carved groove)), do: add_fact(facts, "Has asked about Blood Oghams"), else: facts

    Enum.take(facts, -10)
  end

  @doc "Calculate reputation change from a message."
  def reputation_delta(message) do
    msg = String.downcase(message)
    cond do
      contains_any?(msg, ~w(thank grateful appreciate impressive)) -> Enum.random([3, 4, 5])
      contains_any?(msg, ~w(damn fool idiot useless worthless pathetic)) -> Enum.random([-5, -6, -8])
      contains_any?(msg, ~w(help please need urgent)) -> 1
      true -> 0
    end
  end

  # ══════════════════════════════════════════════════════════════════
  # PROMPT BUILDER
  # ══════════════════════════════════════════════════════════════════

  defp build_prompt(opts) do
    npc = opts[:npc] || %{}
    player = opts[:player] || %{}
    message = opts[:message] || ""
    history = opts[:history] || []
    memory = opts[:memory] || %{}
    _world_flags = opts[:world_flags] || %{}
    _region = opts[:region]
    ai_config = opts[:ai_config] || %{}

    npc_name = npc["name"] || npc[:name] || "Stranger"
    persona = npc["persona"] || npc[:persona] || "A weathered inhabitant of this world."

    hist = history
      |> Enum.take(-8)
      |> Enum.map(fn h ->
        role = h["role"] || h[:role]
        text = h["text"] || h[:text]
        if role == "user", do: "PLAYER: #{text}", else: "#{String.upcase(npc_name)}: #{text}"
      end)
      |> Enum.join("\n")

    facts = memory["facts"] || memory[:facts] || []
    rep = memory["reputation"] || memory[:reputation] || 0

    facts_str = if facts != [] do
      "What you remember about #{player[:name] || "the player"}:\n" <>
        Enum.map_join(facts, "\n", fn f -> "  - #{f}" end)
    else
      ""
    end

    rep_tone = cond do
      rep > 50 -> "You respect and like this player. Be warm and generous with information."
      rep < -30 -> "You distrust this player deeply. Be cold, guarded, or openly hostile."
      true -> "You are neutral toward this player. Cautious but not unfriendly."
    end

    world_desc = (ai_config["systemPrompt"] || ai_config[:system_prompt] || "")
      |> to_string()
      |> String.trim()
      |> then(fn s -> if s == "", do: "a dark Celtic fantasy world. Dark, grounded tone. Celtic gothic.", else: s end)

    address = if player[:title], do: "#{player[:name]} #{player[:title]}", else: player[:name] || "traveler"

    [
      "You are #{npc_name}, a character in #{world_desc}",
      "Your persona: #{persona}",
      "You are speaking to #{address}.",
      rep_tone,
      facts_str,
      "Rules:",
      "- Reply as #{npc_name} ONLY. 1-3 short sentences maximum.",
      "- Stay completely in character. No meta-commentary.",
      "- Do NOT start your reply with your own name.",
      "",
      if(hist != "", do: "Conversation so far:\n#{hist}\n", else: ""),
      "PLAYER: #{message}",
      "#{String.upcase(npc_name)}:"
    ]
    |> Enum.filter(fn s -> s != nil and s != "" end)
    |> Enum.join("\n")
  end

  # ══════════════════════════════════════════════════════════════════
  # AI PROVIDERS
  # ══════════════════════════════════════════════════════════════════

  defp call_provider(provider, prompt, config) do
    api_key = config["apiKey"] || config[:api_key] || ""
    model = config["model"] || config[:model] || ""
    base_url = config["baseUrl"] || config[:base_url] || ""
    max_tokens = config["maxTokens"] || config[:max_tokens] || 256
    temperature = config["temperature"] || config[:temperature] || 0.85

    try do
      case provider do
        :gemini ->
          m = if model != "", do: model, else: "gemini-1.5-flash"
          url = "https://generativelanguage.googleapis.com/v1beta/models/#{m}:generateContent?key=#{api_key}"
          body = Jason.encode!(%{
            contents: [%{parts: [%{text: prompt}]}],
            generationConfig: %{temperature: temperature, maxOutputTokens: max_tokens}
          })
          case http_post(url, body) do
            {:ok, json} ->
              text = get_in(json, ["candidates", Access.at(0), "content", "parts", Access.at(0), "text"])
              if text, do: {:ok, String.trim(text)}, else: nil
            _ -> nil
          end

        :anthropic ->
          url = "https://api.anthropic.com/v1/messages"
          headers = [{"x-api-key", api_key}, {"anthropic-version", "2023-06-01"}, {"content-type", "application/json"}]
          m = if model != "", do: model, else: "claude-haiku-4-5-20251001"
          body = Jason.encode!(%{model: m, max_tokens: max_tokens, messages: [%{role: "user", content: prompt}]})
          case http_post(url, body, headers) do
            {:ok, json} ->
              text = get_in(json, ["content", Access.at(0), "text"])
              if text, do: {:ok, String.trim(text)}, else: nil
            _ -> nil
          end

        :openai ->
          endpoint = if base_url != "", do: String.trim_trailing(base_url, "/") <> "/chat/completions",
                      else: "https://api.openai.com/v1/chat/completions"
          headers = [{"authorization", "Bearer #{api_key}"}, {"content-type", "application/json"}]
          m = if model != "", do: model, else: "gpt-4o-mini"
          body = Jason.encode!(%{model: m, max_tokens: max_tokens, temperature: temperature, messages: [%{role: "user", content: prompt}]})
          case http_post(endpoint, body, headers) do
            {:ok, json} ->
              text = get_in(json, ["choices", Access.at(0), "message", "content"])
              if text, do: {:ok, String.trim(text)}, else: nil
            _ -> nil
          end

        :ollama ->
          url = if base_url != "", do: base_url, else: "http://localhost:11434/api/generate"
          m = if model != "", do: model, else: "llama3"
          body = Jason.encode!(%{model: m, prompt: prompt, stream: false})
          case http_post(url, body) do
            {:ok, json} ->
              text = json["response"] || json["text"]
              if text, do: {:ok, String.trim(to_string(text))}, else: nil
            _ -> nil
          end
      end
    rescue
      e ->
        Logger.warning("NPC Brain #{provider} error: #{Exception.message(e)}")
        nil
    end
  end

  defp http_post(url, body, extra_headers \\ []) do
    headers = [{"content-type", "application/json"} | extra_headers]
    case :httpc.request(:post, {String.to_charlist(url), Enum.map(headers, fn {k, v} -> {String.to_charlist(k), String.to_charlist(v)} end), ~c"application/json", body}, [{:timeout, 15000}], []) do
      {:ok, {{_, 200, _}, _, resp_body}} ->
        case Jason.decode(to_string(resp_body)) do
          {:ok, json} -> {:ok, json}
          _ -> :error
        end
      {:ok, {{_, status, _}, _, _}} ->
        Logger.warning("NPC Brain HTTP #{status}")
        :error
      {:error, reason} ->
        Logger.warning("NPC Brain HTTP error: #{inspect(reason)}")
        :error
    end
  end

  # ══════════════════════════════════════════════════════════════════
  # RULE-BASED FALLBACK
  # ══════════════════════════════════════════════════════════════════

  defp rule_based_reply(npc, message, memory) do
    npc_name = (npc || %{})["name"] || (npc || %{})[:name] || "NPC"
    msg = String.downcase(message || "")
    rep = (memory || %{})["reputation"] || (memory || %{})[:reputation] || 0
    mood = (npc || %{})["mood"] || (npc || %{})[:mood]

    cond do
      rep < -50 -> pick(["\"I have nothing to say to you.\"", "*#{npc_name} turns away without a word.*"])
      mood == "grieving" -> pick(["\"Not now. Please.\"", "*#{npc_name}'s eyes are red.* \"...What do you need?\""])
      mood == "fearful" -> pick(["*#{npc_name} glances around nervously.* \"Keep your voice down.\"", "\"Something's wrong here.\""])
      mood == "angry" -> pick(["\"You picked a bad time.\"", "\"Speak. But choose your words.\""])
      mood == "excited" -> pick(["\"You came at the right time!\"", "\"I was hoping someone would pass through.\""])
      contains_any?(msg, ~w(help danger urgent please)) -> pick(["\"Tell me what happened. Quickly.\"", "\"I'm listening. What's wrong?\""])
      contains_any?(msg, ~w(quest job task mission work)) -> pick(["\"There's always something that needs doing.\"", "\"You want coin or purpose? Choose one.\""])
      contains_any?(msg, ~w(buy sell shop trade)) -> pick(["\"Got something worth trading?\"", "\"Everything has a price.\""])
      contains_any?(msg, ~w(bye goodbye farewell later)) -> pick(["\"Watch your back.\"", "*#{npc_name} nods once.* \"Until next time.\""])
      true -> pick(["\"Careful. That kind of talk gets people noticed.\"", "\"Say what you mean, traveler.\"", "*#{npc_name} studies you for a moment.* \"Ask a clearer question.\""])
    end
  end

  # ══════════════════════════════════════════════════════════════════
  # HELPERS
  # ══════════════════════════════════════════════════════════════════

  def load_ai_config do
    keys = ~w(ai_provider ai_api_key ai_model ai_base_url ai_temperature ai_max_tokens ai_system_prompt)
    config = Enum.reduce(keys, %{}, fn key, acc ->
      case Repo.query("SELECT setting_value FROM system_settings WHERE setting_key=? LIMIT 1", [key]) do
        {:ok, %{rows: [[val]]}} -> Map.put(acc, key, val)
        _ -> acc
      end
    end)

    %{
      "provider" => config["ai_provider"],
      "apiKey" => config["ai_api_key"],
      "model" => config["ai_model"],
      "baseUrl" => config["ai_base_url"],
      "temperature" => parse_float(config["ai_temperature"], 0.85),
      "maxTokens" => parse_int(config["ai_max_tokens"], 256),
      "systemPrompt" => config["ai_system_prompt"]
    }
  end

  defp pick(list), do: Enum.random(list)
  defp contains_any?(str, words), do: Enum.any?(words, fn w -> String.contains?(str, w) end)
  defp add_fact(facts, fact) do
    if fact in facts or length(facts) >= 10, do: facts, else: facts ++ [fact]
  end

  defp parse_int(nil, default), do: default
  defp parse_int(val, _default) when is_integer(val), do: val
  defp parse_int(val, default) when is_binary(val) do
    case Integer.parse(val) do {n, _} -> n; :error -> default end
  end
  defp parse_int(_, default), do: default

  defp parse_float(nil, default), do: default
  defp parse_float(val, _default) when is_float(val), do: val
  defp parse_float(val, default) when is_binary(val) do
    case Float.parse(val) do {f, _} -> f; :error -> default end
  end
  defp parse_float(_, default), do: default
end

File: monorepo.retired/te/npc_brain.js
// =================================================================
// NPC BRAIN v3 — Multi-provider AI with rule-based fallback
// =================================================================
// Provider priority (DB config wins over .env):
//   1. DB setting: ai_provider (gemini | anthropic | openai | ollama | disabled)
//   2. .env fallback: GEMINI_API_KEY → gemini, NPC_LLM_URL → ollama
//   3. Rule-based keyword replies — always works, zero config
//
// To configure: AdminSauce → Settings → 🤖 AI Brain
// Keys are stored in system_settings. The server passes them in as
// aiConfig so this file never touches the DB directly.
// =================================================================

function pick(arr) { return arr[Math.floor(Math.random() * arr.length)]; }
function containsAny(s, words) { return words.some(w => s.toLowerCase().includes(w)); }

// -----------------------------------------------------------------
// GEMINI CALLER
// -----------------------------------------------------------------
async function callGemini(prompt, apiKey, model, temperature, maxTokens) {
    if (!apiKey) throw new Error('No Gemini API key configured');
    const m   = model || 'gemini-1.5-flash';
    const url = `https://generativelanguage.googleapis.com/v1beta/models/${m}:generateContent?key=${apiKey}`;

    const res = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: {
                temperature:     temperature ?? 0.85,
                maxOutputTokens: maxTokens   ?? 256,
                stopSequences:   ['\nPLAYER:', '\n---']
            },
            safetySettings: [
                { category: 'HARM_CATEGORY_HARASSMENT',        threshold: 'BLOCK_NONE' },
                { category: 'HARM_CATEGORY_HATE_SPEECH',       threshold: 'BLOCK_NONE' },
                { category: 'HARM_CATEGORY_SEXUALLY_EXPLICIT', threshold: 'BLOCK_MEDIUM_AND_ABOVE' },
                { category: 'HARM_CATEGORY_DANGEROUS_CONTENT', threshold: 'BLOCK_NONE' }
            ]
        })
    });

    if (res.status === 429) {
        const retryAfter = res.headers.get('Retry-After') || '60';
        throw Object.assign(
            new Error(`Gemini rate limit (429). Retry after ${retryAfter}s.`),
            { isRateLimit: true, retryAfter: parseInt(retryAfter) }
        );
    }
    if (!res.ok) {
        const err = await res.text().catch(() => '');
        throw new Error(`Gemini HTTP ${res.status}: ${err.slice(0, 120)}`);
    }
    const json = await res.json();
    const text = json?.candidates?.[0]?.content?.parts?.[0]?.text;
    if (!text) throw new Error('Gemini returned no text');
    return text.trim();
}

// -----------------------------------------------------------------
// ANTHROPIC (Claude) CALLER
// Uses the /v1/messages endpoint. Model default: claude-haiku (fast + cheap)
// -----------------------------------------------------------------
async function callAnthropic(prompt, apiKey, model, maxTokens, temperature) {
    if (!apiKey) throw new Error('No Anthropic API key configured');
    const res = await fetch('https://api.anthropic.com/v1/messages', {
        method: 'POST',
        headers: {
            'Content-Type':    'application/json',
            'x-api-key':       apiKey,
            'anthropic-version': '2023-06-01'
        },
        body: JSON.stringify({
            model:      model || 'claude-haiku-4-5-20251001',
            max_tokens: maxTokens   ?? 256,
            messages:   [{ role: 'user', content: prompt }]
        })
    });
    if (!res.ok) {
        const err = await res.text().catch(() => '');
        throw new Error(`Anthropic HTTP ${res.status}: ${err.slice(0, 120)}`);
    }
    const json = await res.json();
    const text = json?.content?.[0]?.text;
    if (!text) throw new Error('Anthropic returned no text');
    return text.trim();
}

// -----------------------------------------------------------------
// OPENAI / OPENAI-COMPATIBLE CALLER
// Works with OpenAI, Groq, Together, LM Studio, anything OpenAI-shaped.
// Set ai_base_url to override the endpoint (e.g. http://localhost:1234/v1)
// -----------------------------------------------------------------
async function callOpenAI(prompt, apiKey, model, baseUrl, maxTokens, temperature) {
    const endpoint = (baseUrl || 'https://api.openai.com/v1').replace(/\/$/, '') + '/chat/completions';
    const headers  = { 'Content-Type': 'application/json' };
    if (apiKey) headers['Authorization'] = `Bearer ${apiKey}`;

    const res = await fetch(endpoint, {
        method: 'POST',
        headers,
        body: JSON.stringify({
            model:      model || 'gpt-4o-mini',
            max_tokens: maxTokens   ?? 256,
            temperature: temperature ?? 0.85,
            messages:   [{ role: 'user', content: prompt }]
        })
    });
    if (!res.ok) {
        const err = await res.text().catch(() => '');
        throw new Error(`OpenAI HTTP ${res.status}: ${err.slice(0, 120)}`);
    }
    const json = await res.json();
    const text = json?.choices?.[0]?.message?.content;
    if (!text) throw new Error('OpenAI returned no text');
    return text.trim();
}

// -----------------------------------------------------------------
// OLLAMA CALLER (self-hosted local models)
// -----------------------------------------------------------------
async function callOllama(url, model, prompt) {
    if (!url) throw new Error('No Ollama URL configured');
    const res = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ model, prompt, stream: false })
    });
    if (!res.ok) throw new Error(`Ollama HTTP ${res.status}`);
    const json = await res.json();
    return String(json.response || json.text || '').trim();
}

// -----------------------------------------------------------------
// PROMPT BUILDER
// Builds a rich context prompt from NPC persona, memory, world state.
// aiConfig.systemPrompt lets admins set the world tone from AdminSauce.
// -----------------------------------------------------------------
function buildPrompt({ npc, player, message, history, memory, worldFlags, region, aiConfig }) {
    const hist = (history || []).slice(-8)
        .map(h => (h.role === 'user' ? `PLAYER: ${h.text}` : `${npc.name.toUpperCase()}: ${h.text}`))
        .join('\n');

    const facts   = (memory?.facts?.length)
        ? `What you remember about ${player.name}:\n` + memory.facts.map(f => `  - ${f}`).join('\n')
        : '';
    const rep     = memory?.reputation || 0;
    const repTone = rep > 50  ? 'You respect and like this player. Be warm and generous with information.'
                  : rep < -30 ? 'You distrust this player deeply. Be cold, guarded, or openly hostile.'
                  :             'You are neutral toward this player. Cautious but not unfriendly.';

    const worldKnowledge = worldFlags && Object.keys(worldFlags).length
        ? 'World events you are aware of:\n' +
          Object.entries(worldFlags)
              .filter(([, v]) => v === 'true' || v === '1')
              .map(([k]) => `  - ${k.replace(/_/g, ' ')}`)
              .join('\n')
        : '';

    let regionContext = '';
    if (region) {
        const parts = [];
        if (region.danger_level >= 3) parts.push(`This region is dangerous (danger ${region.danger_level}/5). People are on edge.`);
        if (region.corruption_level >= 2) parts.push(`Dark corruption lingers here (level ${region.corruption_level}/5). The air feels wrong.`);
        if (region.faction_control) parts.push(`The ${region.faction_control} hold sway in this area.`);
        if (region.weather_override && region.weather_override !== 'CLEAR') {
            const wx = { RAIN:'Rain is falling steadily.', STORM:'A fierce storm rages.', FOG:'Thick fog rolls through.',
                BLIZZARD:'A blizzard howls.', BLOOD_MOON:'A blood moon hangs in the sky. Something stirs.' };
            if (wx[region.weather_override]) parts.push(wx[region.weather_override]);
        }
        try {
            const tags = typeof region.active_tags_json === 'string'
                ? JSON.parse(region.active_tags_json) : (region.active_tags_json || []);
            if (tags.includes('siege'))        parts.push('The region is under siege.');
            if (tags.includes('famine'))       parts.push('Famine grips the land.');
            if (tags.includes('undead_surge')) parts.push('The dead walk in numbers tonight.');
            if (tags.includes('festival'))     parts.push('A festival is in full swing nearby.');
        } catch {}
        if (parts.length) regionContext = 'Local conditions:\n' + parts.map(p => `  - ${p}`).join('\n');
    }

    const address   = player.title ? `${player.name} ${player.title}` : player.name;
    const playerCtx = [
        player.level      ? `level ${player.level}` : null,
        player.title      ? `known as "${player.title}"` : null,
        player.questsDone > 0 ? `${player.questsDone} quest(s) completed` : null
    ].filter(Boolean).join(', ');

    const moodLine = npc.mood
        ? `You are currently feeling ${npc.mood}. Let that colour your tone naturally — don't announce it.`
        : null;

    // TEACHING: The admin's custom system prompt from AdminSauce replaces the
    // hardcoded "dark Celtic fantasy world" description, so the game master can
    // set any world tone they want without touching this file.
    const worldDesc = (aiConfig?.systemPrompt || '').trim()
        || 'a dark Celtic fantasy world. Dark, grounded tone. Celtic gothic. Weary, not cheerful.';

    return [
        `You are ${npc.name}, a character in ${worldDesc}`,
        `Your persona: ${npc.persona || 'A weathered inhabitant of this world.'}`,
        `You are speaking to ${address}${playerCtx ? ` (${playerCtx})` : ''}.`,
        moodLine,
        repTone,
        facts          || null,
        worldKnowledge || null,
        regionContext  || null,
        `Rules:`,
        `- Reply as ${npc.name} ONLY. 1-3 short sentences maximum.`,
        `- Stay completely in character. No meta-commentary, no disclaimers.`,
        `- Do NOT start your reply with your own name.`,
        ``,
        hist ? `Conversation so far:\n${hist}\n` : ``,
        `PLAYER: ${message}`,
        `${npc.name.toUpperCase()}:`
    ].filter(v => v !== null).join('\n');
}

// -----------------------------------------------------------------
// MAIN ENTRY POINT
// aiConfig is loaded by server.js from system_settings (DB) and passed
// in here. Falls back to .env if not provided, for backward compat.
// -----------------------------------------------------------------
async function getNpcReply({ npc, player, message, history, memory, worldFlags, region, aiConfig }) {
    const prompt = buildPrompt({ npc, player, message, history, memory, worldFlags, region, aiConfig });

    // Resolve provider: DB config → env vars → disabled
    const provider   = aiConfig?.provider   || (process.env.GEMINI_API_KEY ? 'gemini' : process.env.NPC_LLM_URL ? 'ollama' : null);
    const apiKey     = aiConfig?.apiKey     || process.env.GEMINI_API_KEY || '';
    const model      = aiConfig?.model      || process.env.GEMINI_MODEL   || process.env.NPC_LLM_MODEL || '';
    const baseUrl    = aiConfig?.baseUrl    || process.env.NPC_LLM_URL    || '';
    const temperature = aiConfig?.temperature ?? 0.85;
    const maxTokens  = aiConfig?.maxTokens  ?? 256;

    if (provider === 'gemini') {
        try {
            const out = await callGemini(prompt, apiKey, model, temperature, maxTokens);
            if (out) return out;
        } catch (err) {
            if (err.isRateLimit) {
                console.warn(`[NPC Brain] Gemini rate-limited — falling back. Retry in ${err.retryAfter}s.`);
                global._geminiRateLimit = { at: Date.now(), retryAfter: err.retryAfter };
            } else {
                console.warn(`[NPC Brain] Gemini failed for ${npc.name}: ${err.message}`);
            }
        }
    }

    if (provider === 'anthropic') {
        try {
            const out = await callAnthropic(prompt, apiKey, model, maxTokens, temperature);
            if (out) return out;
        } catch (err) {
            console.warn(`[NPC Brain] Anthropic failed for ${npc.name}: ${err.message}`);
        }
    }

    if (provider === 'openai') {
        try {
            const out = await callOpenAI(prompt, apiKey, model, baseUrl, maxTokens, temperature);
            if (out) return out;
        } catch (err) {
            console.warn(`[NPC Brain] OpenAI failed for ${npc.name}: ${err.message}`);
        }
    }

    if (provider === 'ollama') {
        try {
            const ollamaModel = model || 'llama3';
            const out = await callOllama(baseUrl, ollamaModel, prompt);
            if (out) return out;
        } catch (err) {
            console.warn(`[NPC Brain] Ollama failed for ${npc.name}: ${err.message}`);
        }
    }

    // Rule-based fallback — always works, no AI needed
    return ruleBasedReply({ npc, message, memory });
}

// -----------------------------------------------------------------
// RULE-BASED FALLBACK — mood-aware, no AI required
// -----------------------------------------------------------------
function ruleBasedReply({ npc, message, memory }) {
    const s    = message.toLowerCase();
    const rep  = memory?.reputation || 0;
    const mood = npc.mood || null;

    if (rep < -50) return pick([
        `"I have nothing to say to you."`,
        `*${npc.name} turns away without a word.*`,
        `"Leave. Before I change my mind about being civil."`
    ]);

    if (mood === 'grieving') return pick([
        `"Not now. Please."`,
        `*${npc.name}'s eyes are red. They say nothing for a moment.* "...What do you need?"`,
        `"I'm sorry. I'm not… myself today."`
    ]);
    if (mood === 'fearful') return pick([
        `*${npc.name} glances over their shoulder.* "Keep your voice down."`,
        `"Something's wrong here. I can't explain it, but I feel it."`,
        `"You should leave this place. Both of us should."`
    ]);
    if (mood === 'angry') return pick([
        `"You picked a bad time."`,
        `*${npc.name} looks at you like a problem to be solved.* "Make it quick."`,
        `"Speak. But choose your words."`
    ]);
    if (mood === 'excited') return pick([
        `"You came at the right time! Something's happened—"`,
        `"I was hoping someone would pass through. Listen to this."`,
        `*${npc.name} leans in conspiratorially.* "You won't believe what I just heard."`
    ]);

    if (containsAny(s, ['help', 'danger', 'urgent', 'please', 'save']))
        return pick([`"Tell me what happened. Quickly."`, `"I'm listening. What's wrong?"`, `"Careful now. Start from the beginning."`]);
    if (containsAny(s, ['quest', 'job', 'task', 'mission', 'work']))
        return pick([`"There's always something that needs doing in a place like this."`, `"You want coin or you want purpose? Choose one."`, `"Ask around. The desperate ones always have work."`]);
    if (containsAny(s, ['who are you', 'name', 'what are you']))
        return pick([`"${npc.name}. That's all you need."`, `"I was something else once. Now I'm just ${npc.name}."`, `"Names cost nothing. Remembering them costs more."`]);
    if (containsAny(s, ['buy', 'sell', 'shop', 'trade', 'price', 'coin']))
        return pick([`"Got something worth trading? Let's see it."`, `"Everything has a price. The question is whether you can meet it."`, `"Coin first. Questions after."`]);
    if (containsAny(s, ['enemy', 'monster', 'creature', 'dangerous', 'kill', 'fight']))
        return pick([`"Stay on the road and you'll probably survive."`, `"The things out there don't care about your reasons for being here."`, `"Some of them used to be people. That should bother you more than it does."`]);
    if (containsAny(s, ['map', 'direction', 'where', 'road', 'path', 'north', 'south']))
        return pick([`"East road leads nowhere good. West is your best bet."`, `"I'd draw you a map but they change. The land remembers different things than we do."`, `"Follow the river. Don't stop when the water goes quiet."`]);
    if (containsAny(s, ['thanks', 'thank you', 'grateful', 'appreciate']))
        return rep > 30
            ? pick([`"Take care of yourself out there."`, `"You're one of the good ones. Don't prove me wrong."`, `"Come back if you need anything."`])
            : pick([`"Don't thank me yet."`, `"Save it. We'll see how this plays out."`, `"Mmm."`]);
    if (containsAny(s, ['bye', 'goodbye', 'leave', 'later', 'farewell']))
        return pick([`"Watch your back."`, `"Don't die. It's bad for the rest of us."`, `*${npc.name} nods once.* "Until next time."`]);
    if (containsAny(s, ['ogham', 'rune', 'blood', 'carved', 'groove']))
        return pick([`"Blood Oghams are not toys. Every debt they pay, they carve a new one."`, `"You've seen the grooves in blades? Each mark is a promise to something old."`, `"Those who carry them… change. Slowly. Then all at once."`]);

    return pick([
        `"Careful. That kind of talk gets people noticed."`,
        `"Maybe. Maybe not. Depends what you're willing to risk."`,
        `"Say what you mean, traveler. I don't have all night."`,
        `*${npc.name} studies you for a moment before answering.* "Ask a clearer question."`,
        `"There are things happening here that I won't name in the open."`
    ]);
}

// -----------------------------------------------------------------
// FACT EXTRACTION — learns about player from their messages
// -----------------------------------------------------------------
function extractFacts({ playerMessage, npcName, player, memory }) {
    const facts = (memory?.facts) ? [...memory.facts] : [];
    const msg   = playerMessage.toLowerCase();
    const addFact = (f) => { if (!facts.includes(f) && facts.length < 10) facts.push(f); };

    if (containsAny(msg, ['warrior', 'fighter', 'sword', 'shield']))  addFact('Appears to be a warrior type');
    if (containsAny(msg, ['mage', 'wizard', 'spell', 'magic']))       addFact('Appears to be a mage type');
    if (containsAny(msg, ['rogue', 'thief', 'sneak', 'steal']))       addFact('Appears to be a rogue type');
    if (containsAny(msg, ['quest', 'job', 'task', 'mission']))        addFact('Has asked about quests');
    if (containsAny(msg, ['buy', 'sell', 'shop', 'trade', 'price']))  addFact('Is interested in trading');
    if (containsAny(msg, ['thanks', 'thank you', 'grateful']))        addFact('Was polite and thankful');
    if (containsAny(msg, ['damn', 'fool', 'idiot', 'useless']))       addFact('Was rude');
    if (containsAny(msg, ['help', 'danger', 'urgent', 'please']))     addFact('Seemed to need urgent help');
    if (containsAny(msg, ['ogham', 'blood', 'carved', 'groove']))     addFact('Has asked about Blood Oghams');

    return facts;
}

// -----------------------------------------------------------------
// REPUTATION DELTA
// -----------------------------------------------------------------
function reputationDelta(playerMessage) {
    const msg = playerMessage.toLowerCase();
    if (containsAny(msg, ['thank', 'grateful', 'appreciate', 'well done', 'impressive']))
        return pick([3, 5, 4]);
    if (containsAny(msg, ['damn', 'fool', 'idiot', 'useless', 'worthless', 'pathetic']))
        return pick([-5, -8, -6]);
    if (containsAny(msg, ['help', 'please', 'need', 'urgent']))
        return 1;
    return 0;
}

module.exports = { getNpcReply, extractFacts, reputationDelta };
